Source/WebCore/page/csp/ContentSecurityPolicy.cpp

```diff
@@ -808,13 +831,44 @@ void ContentSecurityPolicy::reportViolation(const String& effectiveViolatedDirec
     violationEventInit.effectiveDirective = effectiveViolatedDirective;
     violationEventInit.originalPolicy = violatedDirectiveList.header();
     violationEventInit.sourceFile = info.sourceFile;
+    // WPT expects this behavior:
+    if (usesReportTo && violationEventInit.sourceFile.isNull())
+        violationEventInit.sourceFile = info.documentURI;
     violationEventInit.disposition = violatedDirectiveList.isReportOnly() ? SecurityPolicyViolationEventDisposition::Report : SecurityPolicyViolationEventDisposition::Enforce;
     violationEventInit.statusCode = httpStatusCode;
     violationEventInit.lineNumber =  info.lineNumber;
     violationEventInit.columnNumber = info.columnNumber;
     violationEventInit.sample = info.sample;
     violationEventInit.bubbles = true;
     violationEventInit.composed = true;
+
+    Vector<String> reportURIs;
+    if (usesReportTo) {
+        if (m_reportingClient) {
```

Comment:
When is this member variable null?

Reply:
CSP running in the WebContent process has a ScriptExecutionContext. CSP running in the Network Process does not, but has a registered ReportingClient that it uses to relay state to the relevant WCP.

One possibility I considered was to have everything implement the ReportingClient interface (Document, Worker, and Clients), and do the right thing depending on context (CSP would register the Document or Worker in WCP, and the NetworkResourceLoader in the NetworkProcess.

That might clean up some of the duplicated code a bit.